File: bench/suite/src/st/reverse-complement/go/reverse-complement.go
// reverse-complement: single-threaded, same algorithm as the Benchmarks Game description.
// Reads FASTA from stdin, writes each sequence's reverse complement, 60 chars per line.
package main

import (
	"bufio"
	"bytes"
	"io"
	"os"
)

const lineLen = 60

var table [256]byte

func init() {
	for i := range table {
		table[i] = byte(i)
	}
	pairs := "AT" + "CG" + "GC" + "TA" + "UA" + "MK" + "RY" + "WW" +
		"SS" + "YR" + "KM" + "VB" + "HD" + "DH" + "BV" + "NN"
	for i := 0; i < len(pairs); i += 2 {
		from, to := pairs[i], pairs[i+1]
		table[from] = to
		table[from+('a'-'A')] = to
	}
}

func writeSeq(out *bufio.Writer, seq []byte) {
	for i, j := 0, len(seq)-1; i < j; i, j = i+1, j-1 {
		seq[i], seq[j] = table[seq[j]], table[seq[i]]
	}
	if len(seq)%2 == 1 {
		mid := len(seq) / 2
		seq[mid] = table[seq[mid]]
	}
	for len(seq) > 0 {
		m := lineLen
		if len(seq) < m {
			m = len(seq)
		}
		out.Write(seq[:m])
		out.WriteByte('\n')
		seq = seq[m:]
	}
}

func main() {
	input, err := io.ReadAll(os.Stdin)
	if err != nil {
		panic(err)
	}
	out := bufio.NewWriterSize(os.Stdout, 1<<16)
	defer out.Flush()
	var seq []byte
	for len(input) > 0 {
		var line []byte
		if i := bytes.IndexByte(input, '\n'); i >= 0 {
			line, input = input[:i], input[i+1:]
		} else {
			line, input = input, nil
		}
		if len(line) == 0 {
			continue
		}
		if line[0] == '>' {
			if len(seq) > 0 {
				writeSeq(out, seq)
				seq = seq[:0]
			}
			out.Write(line)
			out.WriteByte('\n')
		} else {
			seq = append(seq, line...)
		}
	}
	if len(seq) > 0 {
		writeSeq(out, seq)
	}
}

File: bench/suite/src/st/reverse-complement/rust/reverse-complement.rs
// reverse-complement: single-threaded, same algorithm as the Benchmarks Game description.
// Reads FASTA from stdin, writes each sequence's reverse complement, 60 chars per line.
use std::io::{self, Read, Write};

const LINE: usize = 60;

fn complement_table() -> [u8; 256] {
    let mut t = [0u8; 256];
    for (i, v) in t.iter_mut().enumerate() {
        *v = i as u8;
    }
    let pairs: &[(u8, u8)] = &[
        (b'A', b'T'), (b'C', b'G'), (b'G', b'C'), (b'T', b'A'),
        (b'U', b'A'), (b'M', b'K'), (b'R', b'Y'), (b'W', b'W'),
        (b'S', b'S'), (b'Y', b'R'), (b'K', b'M'), (b'V', b'B'),
        (b'H', b'D'), (b'D', b'H'), (b'B', b'V'), (b'N', b'N'),
    ];
    for &(from, to) in pairs {
        t[from as usize] = to;
        t[from.to_ascii_lowercase() as usize] = to;
    }
    t
}

fn write_seq<W: Write>(out: &mut W, seq: &mut Vec<u8>, table: &[u8; 256]) -> io::Result<()> {
    seq.reverse();
    for c in seq.iter_mut() {
        *c = table[*c as usize];
    }
    let mut buf = Vec::with_capacity(seq.len() + seq.len() / LINE + 1);
    for chunk in seq.chunks(LINE) {
        buf.extend_from_slice(chunk);
        buf.push(b'\n');
    }
    out.write_all(&buf)?;
    seq.clear();
    Ok(())
}

fn main() -> io::Result<()> {
    let table = complement_table();
    let mut input = Vec::new();
    io::stdin().lock().read_to_end(&mut input)?;
    let stdout = io::stdout();
    let mut out = io::BufWriter::with_capacity(1 << 16, stdout.lock());
    let mut seq: Vec<u8> = Vec::new();
    for line in input.split(|&c| c == b'\n') {
        if line.is_empty() {
            continue;
        }
        if line[0] == b'>' {
            if !seq.is_empty() {
                write_seq(&mut out, &mut seq, &table)?;
            }
            out.write_all(line)?;
            out.write_all(b"\n")?;
        } else {
            seq.extend_from_slice(line);
        }
    }
    if !seq.is_empty() {
        write_seq(&mut out, &mut seq, &table)?;
    }
    out.flush()
}
